plugins/node/opentelemetry-instrumentation-undici/src/types.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { InstrumentationConfig } from '@opentelemetry/instrumentation';
+import type { Attributes, Span } from '@opentelemetry/api';
+
+export interface UndiciRequest {
+  origin: string;
+  method: string;
+  path: string;
+  /**
+   * Serialized string of headers in the form `name: value\r\n` for v5
+   * Array of strings v6
+   */
+  headers: string | string[];
+  /**
+   * Helper method to add headers (from v6)
+   */
+  addHeader: (name: string, value: string) => void;
+  throwOnError: boolean;
+  completed: boolean;
+  aborted: boolean;
+  idempotent: boolean;
+  contentLength: number | null;
+  contentType: string | null;
+  body: any;
+}
+
+export interface UndiciResponse {
+  headers: Buffer[];
+  statusCode: number;
+}
+
+export interface IgnoreRequestFunction<T = UndiciRequest> {
+  (request: T): boolean;
+}
+export interface CustomAttributesFunction<
+  T = UndiciRequest,
+  Q = UndiciResponse
+> {
+  (span: Span, request: T, response: Q): void;
+}
+
+export interface RequestHookFunction<T = UndiciRequest> {
+  (span: Span, request: T): void;
+}
+
+export interface StartSpanHookFunction<T = UndiciRequest> {
+  (request: T): Attributes;
+}
+
+// This package will instrument HTTP requests made through `undici` or  `fetch` global API
+// so it seems logical to have similar options than the HTTP instrumentation
+export interface UndiciInstrumentationConfig<
+  RequestType = UndiciRequest,
+  ResponseType = UndiciResponse
+> extends InstrumentationConfig {
+  /** Not trace all outgoing requests that matched with custom function */
+  ignoreRequestHook?: IgnoreRequestFunction<RequestType>;
+  /** Function for adding custom attributes after response is handled */
+  applyCustomAttributesOnSpan?: CustomAttributesFunction<
```

Comment:
Update: I actually found a few instrumentations (in Python and .NET) that allow doing similar things. Mostly its intention seems to be to add header values (from either the request or the response) as attributes on the span. 

I'll leave the decision whether to include to include this or not up to the future owners (:slightly_smiling_face:) of this instrumentation. I just wanted to point out that there might possible maintenance implications of adding that functionality to the instrumentation from the get-go. :slightly_smiling_face:

Reply:
You're right @pichlermarc 
The intention was to provide similar functionality as HTTP instrumentation. If you check that instrumentation you'll see the corresponding options.
Ref: https://github.com/open-telemetry/opentelemetry-js/blob/fec6286127129ec0082e5dd080242501f15ea0af/experimental/packages/opentelemetry-instrumentation-http/src/types.ts#L102

For that particular `applyCustomAttributesOnSpan` config I agree it may need maintenance specially when semconv is updated. I'm okay to remove it so let's ask @trentm which also volunteered to own the package.

Reply:
David and I discussed this again. Based on premise that `applyCustomAttributesOnSpan` is mainly about adding header values as attributes and that there is a separate option `headersToSpanAttributes` that supports this use case, my inclination is to start with fewer features. 

Also, FWIW, `headersToSpanAttributes` is amenable to static configuration (in a config file or envvars).

It could be left to a separate issue/PR to make the case to add an `applyCustomAttributesOnSpan` option -- ideally with same real, practical use cases for it.

Reply:
removed!